File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.java
class Solution {
     public int subarraysWithKDistinct(int[] nums, int k) {
        return subarraysWithKSize(nums,k)-subarraysWithKSize(nums,k-1);   
       }
    public int subarraysWithKSize(int[] nums, int k) {
            HashMap<Integer,Integer> hs=new HashMap<>();
            int count=0;
            int n=nums.length;
            int window_start=0;
            int window_end=0;
            while(window_end<n){
                hs.put(nums[window_end],hs.getOrDefault(nums[window_end],0)+1);
            
                while(hs.size()==k+1){
                    hs.put(nums[window_start],hs.get(nums[window_start])-1);
                    if(hs.get(nums[window_start])==0)
                        hs.remove(nums[window_start]);
                    window_start++;
                    
                }
                
                count+=window_end-window_start+1;
                window_end++;

            }
            
            
            return count;
            
        }
}

File: 992-subarrays-with-k-different-integers/992-subarrays-with-k-different-integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        
        def helper(nums, k):
            window_start = 0
            window_end = 0
            count = 0
            hash_set = set()
            hash_map = collections.Counter()

            while window_end < len(nums):
                hash_set.add(nums[window_end])
                hash_map[nums[window_end]] += 1

                while len(hash_set) == k+1:
                    hash_map[nums[window_start]] -= 1
                    if hash_map[nums[window_start]] == 0:
                        hash_set.remove(nums[window_start])
                    window_start += 1

                count += (window_end - window_start + 1)

                window_end += 1
            return count
    
        return helper(nums, k) - helper(nums, k - 1)